File: web/src/app/core/context-parse.ts
// Parsing what the prepass calls send back: the tagged scan sections and the
// `N=Speaker` attribution lines.

import {
  CharacterHint,
  FileContext,
  Gender,
  MAX_IDIOMS,
  MAX_TERMS,
  SceneHint,
  TermHint,
  detectParticipants,
  usableIdioms,
} from './context-pass';

// Closing tag optional so a truncated reply still parses.
const SECTION_RE =
  /<(register|characters|terms|idioms|scenes|notes)>\s*([\s\S]*?)\s*(?=<\/\1>|<(?:register|characters|terms|idioms|scenes|notes)>|$)/gi;
const SCENE_RANGE_RE = /^(\d+)\s*(?:-\s*(\d+))?$/;
const ATTRIB_LINE_RE = /^\s*(\d+)\s*=\s*(.+?)\s*$/;

function stripBullet(line: string): string {
  return line.trim().replace(/^[-*•]\s*/, '').trim();
}

function splitOnce(s: string, sep: string): [string, string] {
  const i = s.indexOf(sep);
  return i < 0 ? [s, ''] : [s.slice(0, i), s.slice(i + sep.length)];
}

// Parse the tagged response. Tolerates whitespace and bullet markers.
export function parseContextResponse(text: string): FileContext {
  const sections: Record<string, string> = {};
  SECTION_RE.lastIndex = 0;
  let m: RegExpExecArray | null;
  while ((m = SECTION_RE.exec(text || '')) !== null) {
    sections[m[1].toLowerCase()] = m[2];
  }

  const register = stripBullet((sections['register'] ?? '').split(/\s+/).join(' '));

  const characters: CharacterHint[] = [];
  for (const raw of (sections['characters'] ?? '').split('\n')) {
    const line = stripBullet(raw);
    if (!line || !line.includes('=>')) continue;
    const [srcPart, restPart] = splitOnce(line, '=>');
    let tgt: string, gender: string;
    if (restPart.includes('|')) {
      const idx = restPart.lastIndexOf('|');
      tgt = restPart.slice(0, idx).trim();
      gender = restPart.slice(idx + 1).trim().toLowerCase();
    } else {
      tgt = restPart.trim();
      gender = 'unknown';
    }
    const g: Gender = gender === 'male' || gender === 'female' ? gender : 'unknown';
    const src = srcPart.trim();
    if (src && tgt) characters.push({ source: src, target: tgt, gender: g });
  }

  const pairs = (section: string): TermHint[] => {
    const out: TermHint[] = [];
    for (const raw of (sections[section] ?? '').split('\n')) {
      const line = stripBullet(raw);
      if (!line || !line.includes('=>')) continue;
      const [srcPart, tgtPart] = splitOnce(line, '=>');
      const src = srcPart.trim();
      const tgt = tgtPart.trim();
      if (src && tgt) out.push({ source: src, target: tgt });
    }
    return out;
  };
  const terms = pairs('terms').slice(0, MAX_TERMS);
  // Pruned before the cap, so dropping a poisoned entry lets a usable one in.
  const idioms = usableIdioms(terms, pairs('idioms')).slice(0, MAX_IDIOMS);

  const scenes: SceneHint[] = [];
  for (const raw of (sections['scenes'] ?? '').split('\n')) {
    const line = stripBullet(raw);
    if (!line || !line.includes('=>')) continue;
    const [rangePart, descPart] = splitOnce(line, '=>');
    const desc = descPart.trim();
    const rm = SCENE_RANGE_RE.exec(rangePart.trim());
    if (!desc || !rm) continue;
    let start = parseInt(rm[1], 10);
    let end = rm[2] ? parseInt(rm[2], 10) : start;
    if (end < start) [start, end] = [end, start];
    scenes.push({
      start, end, description: desc,
      participants: detectParticipants(desc, characters),
      attribution: {},
    });
  }

  const notes: string[] = [];
  for (const raw of (sections['notes'] ?? '').split('\n')) {
    const line = stripBullet(raw);
    if (line) notes.push(line);
  }

  return new FileContext(
    register,
    characters.slice(0, 20),
    terms,
    idioms,
    scenes.slice(0, 40),
    notes.slice(0, 4),
  );
}

export function parseAttributionResponse(
  raw: string, scene: SceneHint, characters: CharacterHint[],
): Record<number, string> {
  const valid = new Set<string>(characters.map((h) => h.source));
  valid.add('unknown');
  const out: Record<number, string> = {};
  for (const line of (raw || '').split('\n')) {
    const m = ATTRIB_LINE_RE.exec(line);
    if (!m) continue;
    const n = parseInt(m[1], 10);
    const name = m[2].trim().replace(/^["']|["']$/g, '');
    if (n >= scene.start && n <= scene.end && valid.has(name)) out[n] = name;
  }
  return out;
}

File: cli/core/context_parse.py
"""Parsing what the prepass calls send back: the tagged scan sections and the
`N=Speaker` attribution lines."""

from __future__ import annotations

import re

from .context_pass import (
    MAX_IDIOMS,
    MAX_TERMS,
    CharacterHint,
    FileContext,
    SceneHint,
    TermHint,
    detect_participants,
    usable_idioms,
)

_TAGS = "register|characters|terms|idioms|scenes|notes"
_SECTION_RE = re.compile(
    rf"<(?P<tag>{_TAGS})>\s*"
    r"(?P<body>.*?)\s*"
    rf"(?=</(?P=tag)>|<(?:{_TAGS})>|\Z)",
    re.I | re.S,
)
_SCENE_RANGE_RE = re.compile(r"^(\d+)\s*(?:-\s*(\d+))?$")
_ATTRIB_LINE_RE = re.compile(r"^\s*(\d+)\s*=\s*(.+?)\s*$")


def _strip_bullet(line: str) -> str:
    return line.strip().lstrip("-*• ").strip()


def _pairs(body: str) -> list[TermHint]:
    """`SOURCE => TARGET` lines, in order, blanks and bullets tolerated."""
    hints: list[TermHint] = []
    for raw in body.splitlines():
        line = _strip_bullet(raw)
        if not line or "=>" not in line:
            continue
        src, tgt = line.split("=>", 1)
        if src.strip() and tgt.strip():
            hints.append(TermHint(src.strip(), tgt.strip()))
    return hints


def parse_context_response(text: str) -> FileContext:
    """Parse the tagged response. Tolerates whitespace and bullet markers."""
    sections = {
        m.group("tag").lower(): m.group("body")
        for m in _SECTION_RE.finditer(text or "")
    }

    register = " ".join(sections.get("register", "").split()).strip().lstrip("-*• ").strip()

    characters: list[CharacterHint] = []
    for line in sections.get("characters", "").splitlines():
        line = _strip_bullet(line)
        if not line or "=>" not in line:
            continue
        src, rest = line.split("=>", 1)
        if "|" in rest:
            tgt, gender = rest.rsplit("|", 1)
            tgt, gender = tgt.strip(), gender.strip().lower()
        else:
            tgt, gender = rest.strip(), "unknown"
        if gender not in ("male", "female", "unknown"):
            gender = "unknown"
        if src.strip() and tgt:
            characters.append(CharacterHint(src.strip(), tgt, gender))

    terms = _pairs(sections.get("terms", ""))[:MAX_TERMS]
    # Dropped before the budget is applied, so a poisoned entry cannot spend
    # one of the fifteen slots a usable idiom could have had.
    idioms = usable_idioms(terms, _pairs(sections.get("idioms", "")))

    scenes: list[SceneHint] = []
    for line in sections.get("scenes", "").splitlines():
        line = _strip_bullet(line)
        if not line or "=>" not in line:
            continue
        rng, desc = line.split("=>", 1)
        m = _SCENE_RANGE_RE.match(rng.strip())
        if not m or not desc.strip():
            continue
        start = int(m.group(1))
        end = int(m.group(2)) if m.group(2) else start
        if end < start:
            start, end = end, start
        scenes.append(SceneHint(
            start=start, end=end, description=desc.strip(),
            participants=detect_participants(desc, characters),
        ))

    notes = [line for raw in sections.get("notes", "").splitlines()
             if (line := _strip_bullet(raw))]

    return FileContext(
        register=register,
        characters=characters[:20],
        terms=terms,
        idioms=idioms[:MAX_IDIOMS],
        scenes=scenes[:40],
        notes=notes[:4],
    )


def parse_attribution_response(
    raw: str, scene: SceneHint, characters: list[CharacterHint],
) -> dict[int, str]:
    """Parse `N=SpeakerName` lines, keeping only in-scene blocks and roster names."""
    valid = {h.source for h in characters} | {"unknown"}
    out: dict[int, str] = {}
    for line in (raw or "").splitlines():
        m = _ATTRIB_LINE_RE.match(line)
        if not m:
            continue
        n = int(m.group(1))
        name = m.group(2).strip().strip('"\'')
        if scene.start <= n <= scene.end and name in valid:
            out[n] = name
    return out
